src/profile-logic/profile-data.js

```diff
@@ -594,38 +595,144 @@ export function filterThreadToRange(
   });
 }
 
-export function getCallNodeFromPath(
-  callNodePath: CallNodePath,
+// --------------- CallNodePath and CallNodeIndex manipulations ---------------
+
+// Returns a list of CallNodeIndex from CallNodePaths. This function uses a map
+// to speed up the look-up process.
+export function getCallNodeIndicesFromPaths(
+  callNodePaths: CallNodePath[],
   callNodeTable: CallNodeTable
+): Array<IndexIntoCallNodeTable | null> {
+  // This is a Map<CallNodePathHash, IndexIntoCallNodeTable>. This map speeds up
+  // the look-up process by caching every CallNodePath we handle which avoids
+  // looking up parents again and again.
+  const cache = new Map();
+  return callNodePaths.map(path =>
+    _getCallNodeIndexFromPathWithCache(path, callNodeTable, cache)
+  );
+}
+
+// Returns a CallNodeIndex from a CallNodePath, using and contributing to the
+// cache parameter.
+function _getCallNodeIndexFromPathWithCache(
+  callNodePath: CallNodePath,
+  callNodeTable: CallNodeTable,
+  cache: Map<string, IndexIntoCallNodeTable>
 ): IndexIntoCallNodeTable | null {
-  let fs = -1;
-  for (let i = 0; i < callNodePath.length; i++) {
-    const func = callNodePath[i];
-    let nextFS = -1;
-    for (
-      let callNodeIndex = fs + 1;
-      callNodeIndex < callNodeTable.length;
-      callNodeIndex++
-    ) {
-      if (
-        callNodeTable.prefix[callNodeIndex] === fs &&
-        callNodeTable.func[callNodeIndex] === func
-      ) {
-        nextFS = callNodeIndex;
-        break;
-      }
+  const hashFullPath = hashPath(callNodePath);
+  const result = cache.get(hashFullPath);
+  if (result !== undefined) {
+    // The cache already has the result for the full path.
+    return result;
+  }
+
+  // This array serves as a map and stores the hashes of callNodePath's
+  // parents to speed up the algorithm. First we'll follow the tree from the
+  // bottom towards the top, pushing hashes as we compute them, and then we'll
+  // move back towards the bottom popping hashes from this array.
+  const sliceHashes = [hashFullPath];
+
+  // Step 1: find whether we already computed the index for one of the path's
+  // parents, starting from the closest parent and looping towards the "top" of
+  // the tree.
+  // If we find it for one of the parents, we'll be able to start at this point
+  // in the following look up.
+  let i = callNodePath.length;
+  let index;
+  while (--i > 0) {
+    // Looking up each parent for this call node, starting from the deepest node.
+    // If we find a parent this makes it possible to start the look up from this location.
+    const subPath = callNodePath.slice(0, i); // subPath = callNodePath[0..i-1]
```

Comment:
The `..` notation here is a bit ambiguous. In rust, this syntax is end-exclusive, so `callNodePath.slice(0, i)` is actually equivalent to `&callNodePath[0..i]`. Not sure what to do here - you could use three periods instead of two, or you could use rust's `..=` syntax, or you could keep the two-period syntax but adjust the end index up by one.

Reply:
ah, I especially wanted to use this abstract syntax to make it clearer, but it's actually less clear :)

I'll find something.

Reply:
I removed all these confusing syntaxes and used a correct .slice instead.